### ISSUE TITLE
Do not throw exception when current Song is nonexistent

### ISSUE THREAD

**Issue by Xerus2000** [sghpjuikit/player#77]
When I start the Player and the current Song in the active Playlist does not exist, I get several exceptions:

```
50:24.312 I [JavaFX Applicat] Player - Next item metadata cache copy failed - content does not correspond to correct item. Loading now...
[00007fb5fc003f98] filesystem access error: cannot open file /home/janek/Daten/Downloads/Titans of the Night/Titans of the Night v1.wav (No such file or directory)
[00007fb5fc003f98] core access error: File reading failed
[00007fb5fc003f98] core access error: VLC could not open the file "/home/janek/Daten/Downloads/Titans of the Night/Titans of the Night v1.wav" (No such file or directory).
[00007fb5b0260f68] core input error: open of `file:///home/janek/Daten/Downloads/Titans%20of%20the%20Night/Titans%20of%20the%20Night%20v1.wav' failed
[00007fb5b0260f68] core input error: Your input can't be opened
[00007fb5b0260f68] core input error: VLC is unable to open the MRL 'file:///home/janek/Daten/Downloads/Titans%20of%20the%20Night/Titans%20of%20the%20Night%20v1.wav'. Check the log for details.
50:24.479 E [ForkJoinPool.co] Ext - Reading metadata failed for file /home/janek/Daten/Downloads/Titans of the Night/Titans of the Night v1.wav
java.io.FileNotFoundException: Unable to find:/home/janek/Daten/Downloads/Titans of the Night/Titans of the Night v1.wav
        at org.jaudiotagger.audio.AudioFileIO.checkFileExists(AudioFileIO.java:441)
        at org.jaudiotagger.audio.AudioFileIO.readFile(AudioFileIO.java:351)
        at org.jaudiotagger.audio.AudioFileIO.read(AudioFileIO.java:196)
        at sp.it.pl.audio.tagging.ExtKt.readAudioFile(Ext.kt:44)
        at sp.it.pl.audio.tagging.Metadata.readArt(Metadata.kt:637)
        at sp.it.pl.audio.tagging.Metadata.getReadCoverOfTag(Metadata.kt:631)
        at sp.it.pl.audio.tagging.Metadata.getCover(Metadata.kt:621)
        at sp.it.pl.audio.tagging.Metadata$getCover$1.invoke(Metadata.kt:624)
        at sp.it.pl.audio.tagging.Metadata$getCover$1.invoke(Metadata.kt:87)
        at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:149)
        at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:109)
        at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:133)
        at sp.it.pl.audio.tagging.Metadata.getCover(Metadata.kt:952)
        at sp.it.pl.gui.infonode.ItemInfo$loadCoverOf$1.get(ItemInfo.kt:69)
        at sp.it.pl.gui.infonode.ItemInfo$loadCoverOf$1.get(ItemInfo.kt:22)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1692)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:283)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1603)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
```

These should be caught since this is a common occurrence and not exceptional. The Items are already marked red, which is sufficient feedback.

**Comment by sghpjuikit** [comment]
You report several exceptions, but you have posted only one.

From the looks of it this the problem is in cover reading, which does not seem to handle missing file, which is a bit odd. I'll look into this. Honestly I do not know how to handle these file checks. They are incredibly expensive in a bulk and not exactly useful. 99% of the time We simply want to silence the exception, but try/catch block isnt the most elegant of the bunch. This problem haunts me with image metadata reading and other stuff as well, where performance implications hold me back from implementing a short workaround.

Anyway, this may open a larger issue with any i/o song operation, because usually there's no file check preceding these. It may be difficult to cover all of the affected cases properly and consistently. Some kind of reader monad doing the try/catch on FileNotFound may end up being necessary.

**Comment by sghpjuikit** [comment]
I'm about to close this as non issue. The reported problem is the presence of logged exception, which is a deliberate best-effort solution.

There are few ways to solve this:
- log the exception higher. This puts too much burden on the consumer of the API. It is absolutely normal to log the exception in the catch block and then recover. User experience in unaffected. Exception is logged, but rightfully so - missing file is an exceptional case.
- put an if File.exists() check before reading the file. This is very bad idea as the underlying library fails reading the file if the file is not there already. The check should only occur once. Then there is performance to consider. Assuming File.exists() takes 1ms, importing a 40k songs would take 40s longer, just so it does not print an exception.
- ignore FileNotFoundException. Honestly - I get shivers from just the idea.

**Comment by sghpjuikit** [comment]
I do agree that attempting to play the file may be a bug though, so I will investigate if it can be avoided.

**Comment by sghpjuikit** [comment]
It is not exceptional in this particular case, but is in general - reading cover is a general purpose call. But I do realize now that there is indeed a problem - why is getCover() being called when the playlist item is invalid?

> And adding a File#exists check as safety before playing a file shouldn't be an issue, since it's only a single call.
Yes, but that is already being done I'm sure. I will double check. Still, the getCover() call originates somewhere else. I'll look into this.

**Comment by Xerus2000** [comment]
> You report several exceptions, but you have posted only one.

There are three errors:
- Player metadata cache copy failed
- VLC errors while trying to play the file
- Reading metada failed

**Comment by sghpjuikit** [comment]
Indeed, I noticed too late (although those are errors, not exceptions).

**Comment by sghpjuikit** [comment]
Does the issue still persist? I can not reproduce this issue. Deleting the played file when restarting the application will not cause any errors to be printed for me. Sometimes I get `Next item metadata cache copy failed`, which is not an error (and I have reworded the message).

**Comment by sghpjuikit** [comment]
Interestingly, there was a problem where some system.out calls were lost (it is normally redirected to sl4j logger) so I could not reproduce the issue at first. Now I did.

I added a item.isCorrupted check before playing the song. I also made the check run in designated IO thread. Problem fixed.

I must note that the GeneralPlayer.play() method remains problematic. It is blocking and even requires synchronization. This will have to be fixed.
